File: core/01-function-call/go/tools.go
// tools.go —— 工具注册表：函数本体 + JSON Schema 配在一起。
// 加新工具只需在 init() 里加一个 register(...) 调用；client/main 都不用改。
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/sashabaranov/go-openai"
	"github.com/sashabaranov/go-openai/jsonschema"
)

type toolFn func(args map[string]any) any

type tool struct {
	def openai.Tool
	fn  toolFn
}

var registry = map[string]tool{}

func register(name, desc string, params jsonschema.Definition, fn toolFn) {
	registry[name] = tool{
		def: openai.Tool{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name:        name,
				Description: desc,
				Parameters:  params,
			},
		},
		fn: fn,
	}
}

// schemas 返回给 LLM 的 tools 字段。
func schemas() []openai.Tool {
	out := make([]openai.Tool, 0, len(registry))
	for _, t := range registry {
		out = append(out, t.def)
	}
	return out
}

// call 执行一次工具调用，返回 JSON 字符串（OpenAI tool message 要的格式）。
func call(name, argsJSON string) string {
	t, ok := registry[name]
	if !ok {
		return errJSON(fmt.Sprintf("unknown tool: %s", name))
	}
	var args map[string]any
	if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
		return errJSON(err.Error())
	}
	b, _ := json.Marshal(t.fn(args))
	return string(b)
}

func errJSON(msg string) string {
	b, _ := json.Marshal(map[string]string{"error": msg})
	return string(b)
}

// ---- 三个示例工具 ----

type product struct {
	ID    int    `json:"id"`
	Name  string `json:"name"`
	Price int    `json:"price"`
}

var products = []product{
	{1, "笔记本电脑", 5999},
	{2, "机械键盘", 599},
	{3, "无线鼠标", 199},
}

func init() {
	register("get_weather", "获取指定城市的天气",
		jsonschema.Definition{
			Type: jsonschema.Object,
			Properties: map[string]jsonschema.Definition{
				"city": {Type: jsonschema.String, Description: "城市名，如：北京"},
				"unit": {Type: jsonschema.String, Enum: []string{"celsius", "fahrenheit"}},
			},
			Required: []string{"city"},
		},
		func(args map[string]any) any {
			db := map[string]struct {
				c    float64
				cond string
			}{"北京": {25, "晴"}, "上海": {28, "多云"}, "深圳": {30, "小雨"}}
			city, _ := args["city"].(string)
			w, ok := db[city]
			if !ok {
				w.c, w.cond = 20, "数据不可用"
			}
			temp := w.c
			if unit, _ := args["unit"].(string); unit == "fahrenheit" {
				temp = w.c*9/5 + 32
			}
			return map[string]any{"city": city, "temperature": temp, "condition": w.cond}
		},
	)

	register("calculate", "执行四则运算",
		jsonschema.Definition{
			Type: jsonschema.Object,
			Properties: map[string]jsonschema.Definition{
				"op": {Type: jsonschema.String, Enum: []string{"add", "sub", "mul", "div"}},
				"a":  {Type: jsonschema.Number},
				"b":  {Type: jsonschema.Number},
			},
			Required: []string{"op", "a", "b"},
		},
		func(args map[string]any) any {
			op, _ := args["op"].(string)
			a, _ := args["a"].(float64)
			b, _ := args["b"].(float64)
			switch op {
			case "add":
				return map[string]any{"result": a + b}
			case "sub":
				return map[string]any{"result": a - b}
			case "mul":
				return map[string]any{"result": a * b}
			case "div":
				if b == 0 {
					return map[string]any{"error": "division by zero"}
				}
				return map[string]any{"result": a / b}
			}
			return map[string]any{"error": "unknown op: " + op}
		},
	)

	// 让 LLM 自己拆"价格 500 以上" -> min_price=500，不要在 Go 里重做 NLP。
	register("search_products", "搜索产品。可按关键词和价格区间过滤。",
		jsonschema.Definition{
			Type: jsonschema.Object,
			Properties: map[string]jsonschema.Definition{
				"query":     {Type: jsonschema.String, Description: "关键词，留空匹配全部"},
				"min_price": {Type: jsonschema.Number},
				"max_price": {Type: jsonschema.Number},
			},
		},
		func(args map[string]any) any {
			query, _ := args["query"].(string)
			minP, _ := args["min_price"].(float64)
			maxP, ok := args["max_price"].(float64)
			if !ok {
				maxP = 1e18
			}
			hits := []product{}
			for _, p := range products {
				if (query == "" || strings.Contains(p.Name, query)) && float64(p.Price) >= minP && float64(p.Price) <= maxP {
					hits = append(hits, p)
				}
			}
			return map[string]any{"count": len(hits), "results": hits}
		},
	)
}

File: core/01-function-call/python/tools.py
"""tools.py —— 工具注册表：函数本体和它的 JSON Schema 配在一起。
新增工具只在这里加一项；client / main / test 都不用改。"""
from __future__ import annotations

import json
from typing import Callable

# name -> (callable, schema)。schema 即 OpenAI function 字段内容。
TOOLS: dict[str, tuple[Callable, dict]] = {}


def tool(schema: dict):
    """装饰器：注册一个 LLM 可调用的工具。"""
    def deco(fn: Callable) -> Callable:
        TOOLS[schema["name"]] = (fn, schema)
        return fn
    return deco


@tool({
    "name": "get_weather",
    "description": "获取指定城市的天气",
    "parameters": {
        "type": "object",
        "properties": {
            "city": {"type": "string", "description": "城市名，如：北京"},
            "unit": {"type": "string", "enum": ["celsius", "fahrenheit"], "default": "celsius"},
        },
        "required": ["city"],
    },
})
def get_weather(city: str, unit: str = "celsius") -> dict:
    db = {"北京": (25, "晴"), "上海": (28, "多云"), "深圳": (30, "小雨")}
    c, cond = db.get(city, (20, "数据不可用"))
    temp = c if unit == "celsius" else round(c * 9 / 5 + 32, 1)
    return {"city": city, "temperature": temp, "unit": unit, "condition": cond}


@tool({
    "name": "calculate",
    "description": "执行四则运算",
    "parameters": {
        "type": "object",
        "properties": {
            "op": {"type": "string", "enum": ["add", "sub", "mul", "div"]},
            "a": {"type": "number"},
            "b": {"type": "number"},
        },
        "required": ["op", "a", "b"],
    },
})
def calculate(op: str, a: float, b: float) -> dict:
    if op == "div" and b == 0:
        return {"error": "division by zero"}
    return {"result": {"add": a + b, "sub": a - b, "mul": a * b, "div": a / b if b else None}[op]}


# 让 LLM 自己拆"价格 500 以上" -> min_price=500，不要在 Python 里重新做 NLP。
@tool({
    "name": "search_products",
    "description": "搜索产品。可按关键词和价格区间过滤。",
    "parameters": {
        "type": "object",
        "properties": {
            "query": {"type": "string", "description": "关键词，留空匹配全部"},
            "min_price": {"type": "number"},
            "max_price": {"type": "number"},
        },
    },
})
def search_products(
    query: str = "",
    min_price: float = 0,
    max_price: float = float("inf"),
) -> dict:
    products = [
        {"id": 1, "name": "笔记本电脑", "price": 5999},
        {"id": 2, "name": "机械键盘", "price": 599},
        {"id": 3, "name": "无线鼠标", "price": 199},
    ]
    hits = [p for p in products if (not query or query in p["name"]) and min_price <= p["price"] <= max_price]
    return {"count": len(hits), "results": hits}


def schemas() -> list[dict]:
    """返回给 LLM 的 tools 字段。"""
    return [{"type": "function", "function": s} for _, s in TOOLS.values()]


def call(name: str, args: dict) -> str:
    """执行一次工具调用，返回 JSON 字符串（OpenAI tool message 要的格式）。"""
    fn, _ = TOOLS.get(name, (None, None))
    if fn is None:
        return json.dumps({"error": f"unknown tool: {name}"}, ensure_ascii=False)
    try:
        return json.dumps(fn(**args), ensure_ascii=False)
    except Exception as e:
        return json.dumps({"error": str(e)}, ensure_ascii=False)
